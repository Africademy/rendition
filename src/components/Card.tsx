import map from 'lodash/map';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, RenditionSystemProps } from '../common-types';

import asRendition from '../asRendition';
import { px } from '../utils';
import Divider from './Divider';
import { Box, Flex } from './Grid';
import Txt from './Txt';

const Wrapper = styled.div<{ minHeight?: string }>`
	border: solid 1px #dfdede;
	border-radius: ${props => px(props.theme.radius || 3)};
	box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.14);
	padding: 20px;

	background-color: #ffffff;
	min-height: ${props => props.minHeight || 'auto'};
`;

const styledDivider = <Divider height={1} color="#dfdede" />;

const Card = ({ title, cta, rows, children, ...props }: InternalCardProps) => {
	const hasHeader = title || cta;

	const Header = hasHeader && (
		<React.Fragment>
			<Flex justifyContent="space-between" mb={14}>
				<Txt bold fontSize="16px" color="#000">
					{title}
				</Txt>
				{cta}
			</Flex>
			{styledDivider}
		</React.Fragment>
	);

	const Rows =
		rows &&
		map(rows, (row, index: number) => (
			<div key={index}>
				{index > 0 && styledDivider}
				{row}
			</div>
		));

	const Body = children && <Box mt={hasHeader ? 14 : 0}>{children}</Box>;

	return (
		<Wrapper {...props}>
			{Header}
			{Rows}
			{Body}
		</Wrapper>
	);
};

export interface InternalCardProps extends DefaultProps {
	title?: string;
	cta?: JSX.Element;
	rows?: JSX.Element[];
	children?: any;
	minHeight?: string;
}

export type CardProps = InternalCardProps & RenditionSystemProps;

export default asRendition<React.FunctionComponent<CardProps>>(Card);
