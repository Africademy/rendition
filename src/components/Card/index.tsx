import map from 'lodash/map';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, RenditionSystemProps } from '../../common-types';

import asRendition from '../../asRendition';
import { DismissableContainer } from '../../internal/DismissableContainer';
import { Box } from '../Box';
import Divider from '../Divider';
import { Flex } from '../Flex';
import Heading from '../Heading';

const Wrapper = styled(DismissableContainer)<WrapperProps>`
	& {
		min-height: ${props => props.minHeight || 'auto'};
		padding: ${props => (props.small ? '24px' : '32px 44px')};
	}
`;

const Card = ({ title, cta, rows, children, ...props }: InternalCardProps) => {
	const hasHeader = title || cta;

	const Header = hasHeader && (
		<React.Fragment>
			<Flex justifyContent="space-between" alignItems="center">
				<Heading.h5 fontSize={2}>{title}</Heading.h5>
				{cta}
			</Flex>
			<Divider my={2} />
		</React.Fragment>
	);

	const Rows =
		rows &&
		map(rows, (row, index: number) => (
			<Box key={index}>
				{index > 0 && <Divider />}
				{row}
			</Box>
		));

	const Body = children && <Box>{children}</Box>;

	return (
		<Wrapper {...props}>
			<Box width="100%">
				{Header}
				{Rows}
				{Body}
			</Box>
		</Wrapper>
	);
};

interface WrapperProps extends DefaultProps {
	minHeight?: string;
	small?: boolean;
}

export interface InternalCardProps extends WrapperProps {
	title?: string;
	cta?: JSX.Element;
	rows?: JSX.Element[];
	children?: any;
}

export type CardProps = InternalCardProps & RenditionSystemProps;

export default asRendition<React.FunctionComponent<CardProps>>(Card);
