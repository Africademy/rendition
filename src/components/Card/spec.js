/* globals expect, test */
import React from 'react'
import renderer from 'react-test-renderer'
import { Card, Provider } from '../../../dist'

test('Search renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Card>Lorem ipsum dolor so ammet</Card>
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
