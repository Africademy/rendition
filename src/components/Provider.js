import * as React from 'react'
import styled, { ThemeProvider } from 'styled-components'
import defaultTheme from '../theme'

const Base = styled.div`
  font-family: ${props => props.theme.font};
`

const Provider = ({ theme, ...props }) => {
  return (
    <ThemeProvider theme={theme || defaultTheme}>
      <Base {...props} />
    </ThemeProvider>
  )
}

export default Provider
