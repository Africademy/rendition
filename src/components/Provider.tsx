import * as React from 'react';
import styled, { StyledFunction, ThemeProvider } from 'styled-components';
import { DefaultProps, Theme } from '../common-types';
import defaultTheme from '../theme';

export interface ThemedBaseProps extends DefaultProps {
	theme?: Theme;
}

const Base = (styled.div as StyledFunction<DefaultProps>)`
  font-family: ${props => props.theme.font};
`;

const Provider = ({ theme, ...props }: ThemedBaseProps) => {
	return (
		<ThemeProvider theme={theme || defaultTheme}>
			<Base {...props} />
		</ThemeProvider>
	);
};

export default Provider;
