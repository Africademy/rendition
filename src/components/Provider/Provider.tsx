import {
	Grommet
} from 'grommet'
import * as React from 'react';

const Provider = ({ theme, ...props }: any) => {
	return (
		<Grommet theme={theme} {...props} />
	);
};

export default Provider;
