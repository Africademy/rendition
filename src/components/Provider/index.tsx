import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';
import { BreakpointProvider } from './BreakpointProvider';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	const providerTheme = merge(defaultTheme, theme);
	return (
		<BreakpointProvider breakpoints={providerTheme.breakpoints}>
			<Base theme={providerTheme} {...props} />
		</BreakpointProvider>
	);
};

export interface ThemedProvider extends DefaultProps {
	theme?: Partial<Theme>;
}

export default Provider;
