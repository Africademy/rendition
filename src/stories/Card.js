import * as React from 'react'
import { storiesOf } from '@storybook/react'
import styled from 'styled-components'
import { Card, Link, Flex, Button, TextWithCopy } from '../'

const Container = styled.div`
  margin: 30px;
  width: 350px;
`

const rows = [
  <div>Lorem Ipsum dolor si amet</div>,
  <Link href='www.resin.io'>Link</Link>,
  <TextWithCopy
    showCopyButton='always'
    copy='This value has been copied to your clipboard!'
    >
    Row with Copy component
  </TextWithCopy>,
  <Flex justify='space-between'>
    <div>Row with</div>
    <div>Flex</div>
  </Flex>,
  <div>Lorem Ipsum dolor si amet</div>
]

storiesOf('Core/Card', module)
  .addDecorator(story => <Container>{story()}</Container>)
  .addWithInfo('Standard', () => {
    return (
      <Card>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam scelerisque
        euismod risus at gravida. Pellentesque a nunc semper, ultrices lacus
        nec, mattis mauris
      </Card>
    )
  })
  .addWithInfo('With Header', () => {
    return (
      <React.Fragment>
        <Card
          title='Card with Button'
          cta={
            <Button
              plaintext
              primary
              onClick={() => window.alert('Action with Button')}
            >
              Update
            </Button>
          }
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam
          scelerisque euismod risus at gravida. Pellentesque a nunc semper,
          ultrices lacus nec, mattis mauris
        </Card>
        <Card
          mt={3}
          title='Card with Link'
          cta={<Link href='https://resin.io'>Link</Link>}
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam
          scelerisque euismod risus at gravida. Pellentesque a nunc semper,
          ultrices lacus nec, mattis mauris
        </Card>
      </React.Fragment>
    )
  })
  .addWithInfo('With rows', () => {
    return (
      <Card
        title='Title'
        rows={rows}
        cta={<Link href='https://resin.io'>Link</Link>}
      />
    )
  })
  .addWithInfo('Title', () => {
    return (
      <Card
        title='Title'
        minHeight={'300px'}
        cta={<Link href='https://resin.io'>Link</Link>}
        >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam scelerisque
        euismod risus at gravida. Pellentesque a nunc semper, ultrices lacus
        nec, mattis mauris
      </Card>
    )
  })
