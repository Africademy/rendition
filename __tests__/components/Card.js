/* globals expect, test */
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import Card from '../../src/components/Card'

test('Search renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Card>Lorem ipsum dolor so ammet</Card>
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
