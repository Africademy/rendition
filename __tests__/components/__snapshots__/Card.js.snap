// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  className="sc-btzYZH jBDBBH StyledGrommet-sc-19lkkz7-0 bBubtL"
>
  <div
    className="sc-iRbamj fcGLpU sc-gPEVay iZWcVf"
  >
    <div
      className="sc-gzVnrw hZnCYl sc-bZQynM kMvugX"
    >
      Lorem ipsum dolor so ammet
    </div>
  </div>
</div>
`;
